File: week-02/cpp/02_constructors.cpp
/**
 * Week 2 - Constructors and Destructors in C++
 */

#include <iostream>
#include <string>
#include <cmath>

using namespace std;

class Resource {
private:
    string name;
    int size;
    static int resourceCount;

public:
    // Constructor
    Resource(string n, int s) : name(n), size(s) {
        resourceCount++;
        cout << "Resource '" << name << "' created (size: " << size << " MB)" << endl;
        cout << "Total resources: " << resourceCount << endl;
    }
    
    // Destructor
    ~Resource() {
        resourceCount--;
        cout << "Resource '" << name << "' destroyed" << endl;
        cout << "Remaining resources: " << resourceCount << endl;
    }
    
    static int getResourceCount() {
        return resourceCount;
    }
};

int Resource::resourceCount = 0;


class Point {
private:
    double x, y;

public:
    // Default constructor
    Point() : x(0), y(0) {
        cout << "Point created at (" << x << ", " << y << ")" << endl;
    }
    
    // Parameterized constructor
    Point(double xVal, double yVal) : x(xVal), y(yVal) {
        cout << "Point created at (" << x << ", " << y << ")" << endl;
    }
    
    // Copy constructor
    Point(const Point& other) : x(other.x), y(other.y) {
        cout << "Point copied to (" << x << ", " << y << ")" << endl;
    }
    
    double distanceFromOrigin() {
        return sqrt(x * x + y * y);
    }
    
    void display() {
        cout << "Point(" << x << ", " << y << ")";
    }
};


class BankAccount {
private:
    string accountNumber;
    string ownerName;
    double balance;
    string accountType;

public:
    // Constructor with default parameters
    BankAccount(string accNum, string owner, double bal = 0.0, string type = "Savings")
        : accountNumber(accNum), ownerName(owner), balance(bal), accountType(type) {
        cout << "Account created: " << accountNumber << " - " << ownerName << endl;
    }
    
    void display() {
        cout << "BankAccount('" << accountNumber << "', '" << ownerName 
             << "', " << balance << ", '" << accountType << "')" << endl;
    }
};


int main() {
    cout << string(60, '=') << endl;
    cout << "Resource Management Example" << endl;
    cout << string(60, '=') << endl;
    
    {
        Resource res1("Image1", 5);
        Resource res2("Video1", 100);
        cout << endl;
        
        // res1 and res2 will be destroyed at end of this block
    }
    
    cout << "\nAfter block - resources destroyed" << endl;
    
    Resource res3("Audio1", 3);
    cout << "\nEnd of program - remaining objects will be destroyed\n" << endl;
    
    cout << string(60, '=') << endl;
    cout << "Point Example" << endl;
    cout << string(60, '=') << endl;
    
    Point p1;  // Default constructor
    Point p2(3, 4);  // Parameterized constructor
    Point p3 = p2;  // Copy constructor
    
    cout << "\np1: ";
    p1.display();
    cout << ", Distance: " << p1.distanceFromOrigin() << endl;
    
    cout << "p2: ";
    p2.display();
    cout << ", Distance: " << p2.distanceFromOrigin() << endl;
    
    cout << "\n" << string(60, '=') << endl;
    cout << "BankAccount Example" << endl;
    cout << string(60, '=') << endl;
    
    // Different ways to create objects
    BankAccount acc1("ACC001", "John Doe");
    BankAccount acc2("ACC002", "Jane Smith", 1000.0);
    BankAccount acc3("ACC003", "Bob Johnson", 5000.0, "Checking");
    
    cout << endl;
    acc1.display();
    acc2.display();
    acc3.display();
    
    return 0;
}

File: week-02/python/02_constructors.py
"""
Week 2 - Constructors and Destructors in Python
"""

class Resource:
    """Demonstrates constructor and destructor"""
    
    resource_count = 0
    
    def __init__(self, name, size):
        """Constructor - called when object is created"""
        self.name = name
        self.size = size
        Resource.resource_count += 1
        print(f"Resource '{self.name}' created (size: {self.size} MB)")
        print(f"Total resources: {Resource.resource_count}")
    
    def __del__(self):
        """Destructor - called when object is destroyed"""
        Resource.resource_count -= 1
        print(f"Resource '{self.name}' destroyed")
        print(f"Remaining resources: {Resource.resource_count}")


class Point:
    """A class representing a 2D point"""
    
    def __init__(self, x=0, y=0):
        """Constructor with default parameters"""
        self.x = x
        self.y = y
        print(f"Point created at ({self.x}, {self.y})")
    
    def distance_from_origin(self):
        """Calculate distance from origin"""
        return (self.x ** 2 + self.y ** 2) ** 0.5
    
    def __str__(self):
        """String representation"""
        return f"Point({self.x}, {self.y})"


class BankAccount:
    """Bank account with multiple constructor options"""
    
    def __init__(self, account_number, owner_name, balance=0.0, account_type="Savings"):
        """
        Constructor with default parameters
        
        Args:
            account_number: Account identifier
            owner_name: Name of account owner
            balance: Initial balance (default 0.0)
            account_type: Type of account (default "Savings")
        """
        self.account_number = account_number
        self.owner_name = owner_name
        self.balance = balance
        self.account_type = account_type
        print(f"Account created: {account_number} - {owner_name}")
    
    def __repr__(self):
        """Developer-friendly representation"""
        return f"BankAccount('{self.account_number}', '{self.owner_name}', {self.balance}, '{self.account_type}')"


if __name__ == "__main__":
    print("=" * 60)
    print("Resource Management Example")
    print("=" * 60)
    
    res1 = Resource("Image1", 5)
    res2 = Resource("Video1", 100)
    print()
    
    # Delete an object explicitly
    del res1
    print()
    
    res3 = Resource("Audio1", 3)
    print("\nEnd of program - remaining objects will be destroyed\n")
    
    print("=" * 60)
    print("Point Example")
    print("=" * 60)
    
    p1 = Point()  # Default constructor
    p2 = Point(3, 4)  # Custom coordinates
    
    print(f"p1: {p1}, Distance: {p1.distance_from_origin():.2f}")
    print(f"p2: {p2}, Distance: {p2.distance_from_origin():.2f}")
    
    print("\n" + "=" * 60)
    print("BankAccount Example")
    print("=" * 60)
    
    # Different ways to create objects
    acc1 = BankAccount("ACC001", "John Doe")
    acc2 = BankAccount("ACC002", "Jane Smith", 1000.0)
    acc3 = BankAccount("ACC003", "Bob Johnson", 5000.0, "Checking")
    
    print(f"\n{repr(acc1)}")
    print(f"{repr(acc2)}")
    print(f"{repr(acc3)}")
